File: app/controllers/ExamMachineController.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package controllers;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.node.ObjectNode;
import controllers.base.BaseController;
import io.ebean.DB;
import io.ebean.Query;
import io.ebean.text.PathProperties;
import java.util.List;
import java.util.Optional;
import models.ExamMachine;
import models.ExamRoom;
import models.Reservation;
import models.Software;
import org.joda.time.DateTime;
import play.libs.Json;
import play.mvc.Http;
import play.mvc.Result;

public class ExamMachineController extends BaseController {

    @Restrict({ @Group("ADMIN") })
    public Result getExamMachines() {
        List<ExamMachine> machines = DB.find(ExamMachine.class).where().eq("archived", false).findList();

        return ok(machines);
    }

    @Restrict({ @Group("ADMIN"), @Group("STUDENT") })
    public Result getExamMachine(Long id) {
        PathProperties pp = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))");
        Query<ExamMachine> query = DB.find(ExamMachine.class);
        pp.apply(query);
        ExamMachine machine = query.where().idEq(id).findOne();
        return ok(machine, pp);
    }

    @Restrict({ @Group("ADMIN") })
    public Result getExamMachineReservationsFromNow(Long id) {
        List<Reservation> reservations = DB
            .find(Reservation.class)
            .where()
            .eq("machine.id", id)
            .gt("endAt", DateTime.now()) //.minus(1000 * 60 * 60 * 24))
            .findList();

        return ok(reservations);
    }

    @Restrict({ @Group("ADMIN") })
    public Result updateExamMachine(Long id, Http.Request request) {
        ExamMachine src = formFactory
            .form(ExamMachine.class)
            .bindFromRequest(
                request,
                "id",
                "name",
                "otherIdentifier",
                "accessibilityInfo",
                "accessible",
                "ipAddress",
                "surveillanceCamera",
                "videoRecordings",
                "expanded",
                "statusComment",
                "outOfService"
            )
            .get();
        ExamMachine dest = DB.find(ExamMachine.class, id);
        if (dest == null) {
            return notFound();
        }

        if (!src.getIpAddress().isEmpty()) {
            List<ExamMachine> machines = DB.find(ExamMachine.class).findList();
            List<String> ips = machines.stream().filter(m -> !m.equals(dest)).map(ExamMachine::getIpAddress).toList();
            if (ips.contains(src.getIpAddress())) {
                return forbidden("i18n_error_ip_address_exists_for_room");
            }
        }

        dest.setName(src.getName());
        dest.setOtherIdentifier(src.getOtherIdentifier());
        dest.setAccessibilityInfo(src.getAccessibilityInfo());
        dest.setAccessible(src.isAccessible());
        dest.setIpAddress(src.getIpAddress());
        dest.setSurveillanceCamera(src.getSurveillanceCamera());
        dest.setVideoRecordings(src.getVideoRecordings());
        dest.setExpanded(src.getExpanded());
        dest.setStatusComment(src.getStatusComment());
        dest.setOutOfService(src.getOutOfService());

        dest.update();
        PathProperties pp = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))");

        return ok(dest, pp);
    }

    @Restrict({ @Group("ADMIN") })
    public Result resetMachineSoftware(Long mid) {
        ExamMachine machine = DB.find(ExamMachine.class, mid);
        if (machine == null) {
            return notFound();
        }

        machine.getSoftwareInfo().clear();
        machine.update();
        PathProperties pp = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))");

        return ok(machine, pp);
    }

    @Restrict({ @Group("ADMIN") })
    public Result updateMachineSoftware(Long mid, Long sid) {
        ExamMachine machine = DB.find(ExamMachine.class, mid);
        if (machine == null) {
            return notFound();
        }
        boolean isTurnedOn = false;
        Software software = DB.find(Software.class, sid);
        if (machine.getSoftwareInfo().contains(software)) {
            machine.getSoftwareInfo().remove(software);
        } else {
            machine.getSoftwareInfo().add(software);
            isTurnedOn = true;
        }
        machine.update();
        ObjectNode part = Json.newObject();
        part.put("turnedOn", isTurnedOn);

        return ok(Json.toJson(part));
    }

    @Restrict({ @Group("ADMIN") })
    public Result insertExamMachine(Long id) {
        var room = DB.find(ExamRoom.class, id);
        if (room == null) {
            return notFound();
        }
        var machine = new ExamMachine();
        room.getExamMachines().add(machine);
        room.save();
        machine.save();
        return ok(machine);
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result removeExamMachine(Long id) {
        // SIT-690
        return forbidden();
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN"), @Group("STUDENT") })
    public Result listSoftware() {
        List<Software> software = DB
            .find(Software.class)
            .where()
            .or()
            .isNull("status")
            .eq("status", "ACTIVE")
            .endOr()
            .findList();

        return ok(software);
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result getSoftware(Long id) {
        Software software = DB.find(Software.class, id);

        return ok(software);
    }

    private Optional<Result> checkSoftwareName(String name) {
        List<Software> sw = DB.find(Software.class).where().ieq("name", name).findList();
        return sw.isEmpty() ? Optional.empty() : Optional.of(badRequest("Software with that name already exists"));
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result addSoftware(String name, Http.Request request) {
        return checkSoftwareName(name)
            .orElseGet(() -> {
                Software software = new Software();
                software.setName(name);
                software.save();
                return ok(software);
            });
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result updateSoftware(Long id, String name) {
        Software software = DB.find(Software.class, id);
        if (software == null) {
            return notFound();
        }
        return checkSoftwareName(name)
            .orElseGet(() -> {
                software.setName(name);
                software.update();
                return ok(software);
            });
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result removeSoftware(Long id) {
        Software software = DB.find(Software.class, id);
        if (software == null) {
            return notFound();
        }
        software.delete();
        return ok();
    }
}

File: app/controllers/facility/ExamMachineController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.facility

import io.ebean.DB
import io.ebean.text.PathProperties
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.enrolment.Reservation
import models.facility.{ExamMachine, ExamRoom, Software}
import models.user.Role
import org.joda.time.DateTime
import play.api.libs.json.{JsValue, Json}
import play.api.mvc.{Action, AnyContent, BaseController, ControllerComponents}
import security.scala.Auth.{AuthenticatedAction, authorized}

import javax.inject.Inject
import scala.concurrent.ExecutionContext

class ExamMachineController @Inject() (
    authenticated: AuthenticatedAction,
    val controllerComponents: ControllerComponents
)(implicit ec: ExecutionContext)
    extends BaseController
    with DbApiHelper
    with JavaApiHelper:

  def getExamMachines: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val machines = DB.find(classOf[ExamMachine]).where().eq("archived", false).list
      Ok(machines.asJson)
    }

  def getExamMachine(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.STUDENT))) { _ =>
      val pp    = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))")
      val query = DB.find(classOf[ExamMachine])
      pp.apply(query)
      Option(query.where().idEq(id).findOne()) match
        case None          => NotFound("machine not found")
        case Some(machine) => Ok(machine.asJson(pp))
    }

  def getExamMachineReservationsFromNow(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val reservations = DB
        .find(classOf[Reservation])
        .where()
        .eq("machine.id", id)
        .gt("endAt", DateTime.now())
        .list

      Ok(reservations.asJson)
    }

  def updateExamMachine(id: Long): Action[JsValue] =
    authenticated(parse.json).andThen(authorized(Seq(Role.Name.ADMIN))) { request =>
      Option(DB.find(classOf[ExamMachine], id)) match
        case None => NotFound("machine not found")
        case Some(dest) =>
          val body = request.body

          // Check for IP address conflicts
          val ipConflict = (body \ "ipAddress").asOpt[String].filter(_.nonEmpty).flatMap { newIp =>
            val existingIps = DB
              .find(classOf[ExamMachine])
              .list
              .filter(m => !m.equals(dest))
              .flatMap(m => Option(m.getIpAddress))

            if existingIps.contains(newIp) then Some("i18n_error_ip_address_exists_for_room")
            else None
          }

          ipConflict match
            case Some(error) => Forbidden(error)
            case None =>
              (body \ "name").asOpt[String].foreach(dest.setName)
              (body \ "otherIdentifier").asOpt[String].foreach(dest.setOtherIdentifier)
              (body \ "accessibilityInfo").asOpt[String].foreach(dest.setAccessibilityInfo)
              (body \ "accessible").asOpt[Boolean].foreach(v => dest.setAccessible(v))
              (body \ "ipAddress").asOpt[String].foreach(dest.setIpAddress)
              (body \ "surveillanceCamera").asOpt[String].foreach(dest.setSurveillanceCamera)
              (body \ "videoRecordings").asOpt[String].foreach(dest.setVideoRecordings)
              (body \ "expanded").asOpt[Boolean].foreach(v => dest.setExpanded(v))
              (body \ "statusComment").asOpt[String].foreach(dest.setStatusComment)
              (body \ "outOfService").asOpt[Boolean].foreach(v => dest.setOutOfService(v))

              dest.update()
              val pp = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))")
              Ok(dest.asJson(pp))
    }

  def resetMachineSoftware(mid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      Option(DB.find(classOf[ExamMachine], mid)) match
        case None => NotFound("machine not found")
        case Some(machine) =>
          machine.getSoftwareInfo.clear()
          machine.update()
          val pp = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))")
          Ok(machine.asJson(pp))
    }

  def updateMachineSoftware(mid: Long, sid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      Option(DB.find(classOf[ExamMachine], mid)) match
        case None => NotFound("machine not found")
        case Some(machine) =>
          Option(DB.find(classOf[Software], sid)) match
            case None => NotFound("software not found")
            case Some(software) =>
              val isTurnedOn =
                if machine.getSoftwareInfo.contains(software) then
                  machine.getSoftwareInfo.remove(software)
                  false
                else
                  machine.getSoftwareInfo.add(software)
                  true

              machine.update()
              Ok(Json.obj("turnedOn" -> isTurnedOn))
    }

  def insertExamMachine(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      Option(DB.find(classOf[ExamRoom], id)) match
        case None => NotFound("room not found")
        case Some(room) =>
          val machine = new ExamMachine()
          room.getExamMachines.add(machine)
          room.save()
          machine.save()
          Ok(machine.asJson)
    }

  def removeExamMachine(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      // SIT-690
      Forbidden("Machine removal is not allowed")
    }

  def listSoftware: Action[AnyContent] =
    authenticated { _ =>
      val software = DB
        .find(classOf[Software])
        .where()
        .or()
        .isNull("status")
        .eq("status", "ACTIVE")
        .endOr()
        .list

      Ok(software.asJson)
    }

  def getSoftware(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) { _ =>
      Option(DB.find(classOf[Software], id)) match
        case None           => NotFound("software not found")
        case Some(software) => Ok(software.asJson)
    }

  private def checkSoftwareName(name: String): Option[String] =
    val existing = DB.find(classOf[Software]).where().ieq("name", name).list
    if existing.isEmpty then None
    else Some("Software with that name already exists")

  def addSoftware(name: String): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      checkSoftwareName(name) match
        case Some(error) => BadRequest(error)
        case None =>
          val software = new Software()
          software.setName(name)
          software.save()
          Ok(software.asJson)
    }

  def updateSoftware(id: Long, name: String): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      Option(DB.find(classOf[Software], id)) match
        case None => NotFound("software not found")
        case Some(software) =>
          checkSoftwareName(name) match
            case Some(error) => BadRequest(error)
            case None =>
              software.setName(name)
              software.update()
              Ok(software.asJson)
    }

  def removeSoftware(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      Option(DB.find(classOf[Software], id)) match
        case None => NotFound("software not found")
        case Some(software) =>
          software.delete()
          Ok
    }
